### PR TITLE
Functional JUnit reporter

### LINKED ISSUE #1438: jUnit reports are hard to read
### Description

The jUnit exporter is organizing tests under invalid control names. This makes it really hard to read when importing tests results into test reporting tools.

### InSpec and Platform Version

InSpec version: ``1.9.0``
Platform version: ``RedHat 7.2``

### Replication Case

1. Import a control from another profile

audit_config_spec.rb:

```ruby
title 'Audit Configuration Ansible'

require_controls 'STIG_RHEL7' do
  control 'RHEL-07-030090'
  control 'RHEL-07-030310'
  control 'RHEL-07-030380'
  control 'RHEL-07-030381'
  control 'RHEL-07-030382'
  control 'RHEL-07-030383'
  control 'RHEL-07-030390'
  control 'RHEL-07-030391'
  control 'RHEL-07-030392'
  control 'RHEL-07-030400'
  control 'RHEL-07-030401'
  control 'RHEL-07-030402'
  control 'RHEL-07-030403'
  control 'RHEL-07-030404'
  control 'RHEL-07-030405'
  control 'RHEL-07-030420'
  control 'RHEL-07-030421'
  control 'RHEL-07-030422'
  control 'RHEL-07-030423'
  control 'RHEL-07-030424'
  control 'RHEL-07-030425'
  control 'RHEL-07-030441'
  control 'RHEL-07-030442'
  control 'RHEL-07-030443'
  control 'RHEL-07-030444'
  control 'RHEL-07-030490'
  control 'RHEL-07-030491'
  control 'RHEL-07-030492'
  control 'RHEL-07-030510'
  control 'RHEL-07-030511'
  control 'RHEL-07-030512'
  control 'RHEL-07-030513'
  control 'RHEL-07-030514'
  control 'RHEL-07-030521'
  control 'RHEL-07-030522'
  control 'RHEL-07-030523'
  control 'RHEL-07-030524'
  control 'RHEL-07-030525'
  control 'RHEL-07-030526'
  control 'RHEL-07-030530'
  control 'RHEL-07-030531'
  control 'RHEL-07-030540'
  control 'RHEL-07-030541'
  control 'RHEL-07-030550'
  control 'RHEL-07-030560'
  control 'RHEL-07-030561'
  control 'RHEL-07-030630'
  control 'RHEL-07-030670'
  control 'RHEL-07-030671'
  control 'RHEL-07-030672'
  control 'RHEL-07-030673'
  control 'RHEL-07-030674'
  control 'RHEL-07-030710'
  control 'RHEL-07-030750'
  control 'RHEL-07-030751'
  control 'RHEL-07-030752'
  control 'RHEL-07-030753'
  control 'RHEL-07-030754'
end

control 'audit-config-ansible-01' do
  impact 1.0
  title 'Verify audit daemon is configured by Ansible correctly'
  desc 'Audit daemon config should be have a log format of RAW, no have a name set and max log file action should be rotate. Audisp daemon should have q depth set to 150. Audisp remote should not have kerb enabled and not have a remote server set.'

  describe file('/etc/audit/auditd.conf') do
    its('content') { should match /log_format = RAW/ }
    its('content') { should_not match /name =/ }
    its('content') { should match /max_log_file_action = ROTATE/ }
  end

  describe file('/etc/audisp/audispd.conf') do
    its('content') { should match /q_depth = 150/ }
  end

  describe file('/etc/audisp/audisp-remote.conf') do
    its('content') { should match /enable_krb5 = no/ }
    its('content') { should_not match /remote_server = \w+/ }
  end
end
```

2. Generate a junit report

jUnit report: [junit_report.txt](https://github.com/chef/inspec/files/736148/junit_report.txt)
InSpec report: [inspec_report.txt](https://github.com/chef/inspec/files/736149/inspec_report.txt)

3. Import into Jenkins or another jUnit tool
4. Imported or required tests are not all showing up under the same parent 

Main test results page:
screen shot 2017-01-27 at 2 41 02 pm

tmp..inspec.cache.91e12bdaec14f85c7bb7bf2d7c6177ea821175ef.controls page:
screen shot 2017-01-27 at 2 48 34 pm

As you can see a bunch of controls are missing. They seem to appear under usr.lib64.ruby.gems.2.3.0.gems.inspec-1.9.0.lib.inspec:

screen shot 2017-01-27 at 3 06 13 pm

### Possible Solutions

Not sure yet I will take a look at the code when I have a chance.

### Stacktrace

Not applicable

### PR DESCRIPTION
**Pull request by jkerry**
resolves #1438 by removing the gem based junit reporter and manually rolling up JUnit formatted XML based on the JSON reporter's @output_hash.

Some minor ancillary changes:

- The testsuite is no longer expected to have an errors count
- nokogiri gem elevated to a core dependency

### CODE REVIEW

**Comment by jkerry** [comment]
Is there any SOP for digging deeper here on the travis build?  not seeing the same failure when running functional tests on the macbook

**Comment by jkerry** [comment]
no change, rerun, different failure 👎

**Comment by jkerry** [comment]
image

File: lib/inspec/rspec_json_formatter.rb [hunk 1]
```diff
@@ -748,13 +748,73 @@ def status_type(example)
   end
 end
 
-class InspecRspecJUnit < RSpecJUnitFormatter
+class InspecRspecJUnit < InspecRspecJson
```
Comment (chris-rock) [inline lib/inspec/rspec_json_formatter.rb:751]: It is so good to see that this is now based on our InSpec JSON generator. Awesome @jkerry

**Review by chris-rock — APPROVED** [review]
Awesome improvement @jkerry Could you please rebase on latest master?

File: lib/inspec/rspec_json_formatter.rb [hunk 2]
```diff
@@ -748,13 +748,73 @@ def status_type(example)
...
+
+  def build_control_xml(xml, control)
+    return if control[:results].nil?
+    control[:results].each do |result|
+      build_result_xml(xml, control, result)
+    end
+  end
+
+  def build_result_xml(xml, control, result)
+    test_class = control[:title].nil? ? 'Anonymnous' : control[:id]
```
Comment (adamleff) [inline lib/inspec/rspec_json_formatter.rb]: "Anonymous" is misspelled here.

**Review by adamleff — APPROVED** [review]
Small misspelling. otherwise, approved :)

**Comment by jkerry** [comment]
Added!  one peculiarity I noticed while developing this is if a case is explicitly skipped with a guard it won't show up in the json output as 'skipped'. It is simply omitted from the output.  Is that a feature or a bug?  A case that describes, say, a file that isn't present shows in the output as skipped.

**Comment by jkerry** [comment]
Doh, rebase busted the tests.  I'll get this handled today

**Comment by chris-rock** [comment]
@jkerry That may be related to https://github.com/chef/inspec/issues/1515

**Comment by adamleff** [comment]
@chris-rock I don't believe it's related to the rb-readline issues because we'd be seeing a bunch of `inspec_shell_test` failures here too. These look like legitimate test failures, especially this one:

```
  2) Error:
inspec exec with junit formatter::execute a profile with junit formatting::the test suite::the testcase named "gordon_config Can't find file ..."#test_0002_should be skipped:
NoMethodError: undefined method `xpath' for nil:NilClass
    /home/travis/build/chef/inspec/test/functional/inspec_exec_junit_test.rb:71:in `block (5 levels) in <top (required)>'
```

**Comment by jkerry** [comment]
Yeah, this was on me.  I ran a rebase and force push off of a stale local repo. I can't run functional tests on a windows development box so the real source of truth was on the macbook.

**Comment by jkerry** [comment]
@chris-rock @adamleff  technically this was a history rewrite so you might want to revisit the code review just in case

**Comment by chris-rock** [comment]
Thank you for the additional effort @jkerry

**Comment by adamleff** [comment]
Thank you, @jkerry - this looks great. We greatly appreciate your contribution!

**Comment by ryancurrah** [comment]
Woot! @jkerry thanks so much!

**Pull request merged**